sway/commands/swap.c

```diff
@@ -11,7 +13,162 @@
 #include "stringop.h"
 
 static const char expected_syntax[] =
-	"Expected 'swap container with id|con_id|mark <arg>'";
+	"Expected 'swap (container with id|con_id|mark <arg>) |\n\t(wokspace with [number] <name>)'";
+
+/*
+ * prepare the containers inside a workspace to be moved to another
+ * workspace. This includes setting a few properties of the containers
+ * as well as ipc events
+ */
+
+static void handle_container_after_move(struct sway_container *container,
+		void *data) {
+	node_set_dirty(&container->node);
+	container_update_representation(container);
+
+	struct sway_workspace *old_workspace = container->pending.workspace;
+	struct sway_output *old_output = container->pending.workspace->output;
+
+	struct sway_workspace *destination = data;
+	container->pending.workspace = destination;
+
+	// handle floating containers here by updating their position
+	if (container_is_floating(container)) {
+		if (old_output == destination->output ||
+				container->pending.fullscreen_mode) return;
+
+		struct wlr_box workspace_box, old_workspace_box;
+
+		workspace_get_box(destination, &workspace_box);
+		workspace_get_box(old_workspace, &old_workspace_box);
+
+		floating_fix_coordinates(container,
+				&old_workspace_box, &workspace_box);
+
+		if (!container->scratchpad || !destination->output) return;
+
+		struct wlr_box output_box;
+		output_get_box(destination->output, &output_box);
+		container->transform = workspace_box;
+	}
+
+	if (container->view) {
+		ipc_event_window(container, "move");
+	}
+}
+
+
+/*
+ * swap the properties necessary to preserve the layout
+ * as well as their respective contents
+ */
+
+static void swap_workspace_properties(struct sway_workspace *cur_ws,
+		struct sway_workspace *oth_ws) {
+	struct sway_workspace cur_ws_cpy = *cur_ws;
+
+	cur_ws->tiling = oth_ws->tiling;
+	oth_ws->tiling = cur_ws_cpy.tiling;
+	cur_ws->floating = oth_ws->floating;
+	oth_ws->floating = cur_ws_cpy.floating;
+	cur_ws->layout = oth_ws->layout;
+	oth_ws->layout = cur_ws_cpy.layout;
+	cur_ws->prev_split_layout = oth_ws->prev_split_layout;
+	oth_ws->prev_split_layout = cur_ws_cpy.prev_split_layout;
+	cur_ws->current_gaps = oth_ws->current_gaps;
+	oth_ws->current_gaps = cur_ws_cpy.current_gaps;
+	cur_ws->gaps_outer = oth_ws->gaps_outer;
+	oth_ws->gaps_outer = cur_ws_cpy.gaps_outer;
+	cur_ws->gaps_inner = oth_ws->gaps_inner;
+	oth_ws->gaps_inner = cur_ws_cpy.gaps_inner;
+}
+
+static void set_new_focus(struct sway_workspace *ws, struct sway_seat *seat) {
+	if (ws->tiling->length) {
+		// this needs to be more specific (focus not just every container,
+		// but single windows
+		struct sway_container *container = ws->tiling->items[0];
+		struct sway_container *first_view = container_get_first_view(container);
```

Comment:
Why focus first view? Shouldn't `workspace->focused_inactive_child` be the preferred  way?